fix #2258: feedback messages should not be sent if user has already seen to them.

prasanna08 opened the pull request:
When feedback is already seen, feedback message email should not be sent. This commit takes care of that by removing corresponding feedback message references from UnsentFeedbackEmailModel instance so that they are not present in the sent email.

seanlip commented on core/controllers/feedback.py:
```diff
@@ -32,6 +32,8 @@ def get(self, exploration_id):
         self.values.update({
             'threads': [t.to_dict() for t in feedback_services.get_all_threads(
                 exploration_id, False)]})
+        feedback_services.update_feedback_message_references(
```
clear_feedback_message_references might be a better name? "Update" is a bit misleading...

Also, I feel like there might be weird coupling here -- it's possible that the editor page is opened and the feedback threads are loaded, but the user doesn't actually navigate to the feedback tab to see them. Maybe we should send a POST request if the user navigates to the feedback tab, to a separate handler, which is responsible for clearing the queue? That also preserves the general rule that GET handlers should not be state-changing.

seanlip commented on core/domain/email_manager.py:
```diff
@@ -391,9 +391,6 @@ def send_feedback_message_email(recipient_id, feedback_messages):
         log_new_error('This app cannot send feedback message emails to users.')
         return
 
-    if not feedback_messages:
```
Might as well keep the guard in here, I think?

prasanna08 replied:
I don't think that in normal circumstances this will be needed. Since we are already returning if references are empty. And messages will be empty only if references are empty. So empty references will never make it to `send_feedback_message_email`.
Anyway having a one more if condition will not hurt anything.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -294,9 +294,10 @@ def get_feedback_message_references(user_id):
+    model = feedback_models.UnsentFeedbackEmailModel.get(user_id, strict=False)
 
     if model is None:
+        # Model may not exist is user has already attended to feedback.
```
is --> if

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,31 @@ def pop_feedback_message_references(user_id, references_length):
+def update_feedback_message_references(user_id, exploration_id):
+    """Removes feedback message references associated with exploration."""
+    if not user_id:
+        # If user is not logged in this will be True.
```
I think the new controller should require a logged-in user.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,31 @@ def pop_feedback_message_references(user_id, references_length):
+    updated_references = []
+    for reference in model.feedback_message_references:
+        if exploration_id not in reference.values():
+            updated_references.append(reference)
```
Actually, we could go more fine-grained here, and fire an event to the backend when a particular message thread is opened by a user. The event contains the last message in that thread. Then we clear just the messages from that message thread, up to and including the last message. Would this make better sense from a user perspective?

Basically, if the user opens one thread, that's not a guarantee that they've opened the other threads with new messages.

prasanna08 replied:
Ah, so you mean that we should only clear those thread messages which are opened by the user and keeping the rest. Yes, from user perspective this is more helpful  instead how it is implemented right now. As of now all messages get deleted without any regard that they have been seen or not. Deleting only opened threads makes more sense.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,31 @@ def pop_feedback_message_references(user_id, references_length):
+    if not updated_references:
+        # Allow task to process even if model is deleted. Email will not be
+        # sent if model does not exist.
+        model.key.delete()
```
model.delete() (since model subclasses base_models.BaseModel and there's a delete() method already defined there)

seanlip commented:
Took a first pass. Thanks!

prasanna08 commented:
Hi Sean, 
I have added a new backend url as per your suggestion, for clear up task of feedback messages. Also this url will be fired only when a particular thread is opened or reloaded. So now only those messages will be cleared up whose thread is opened by the creator.

Also note this scenario. Consider that exploration got a feedback. This triggered a backend task for sending email to the creator with countdown of an hour. But creator visited exploration and attended to feedback within half hour. So now we deleted that feedback message model and just after he left exploration got a new feedback. So we again created the new model and enqueue a new task for sending email. So now we have 2 emails task running in backend but first task that was triggered by 1st feedback will now execute first. It will find and model will exists and send email for that feedback. (note the time, it sent 2nd feedback in half hour) and now the task triggered by 2nd feedback will execute and find no model exists. So no email will be sent.

I don't think above scenario has any problem or something but just wanted to let you know that this can happen.

Thanks

seanlip commented on core/controllers/feedback.py:
```diff
@@ -244,3 +248,16 @@ def post(self):
+
+
+class FeedbackMessageClear(base.BaseHandler):
+    """Task for clearing feedback messages from email."""
```
... when a feedback thread is viewed by the potential email recipient.

seanlip commented on core/controllers/feedback.py:
```diff
@@ -244,3 +248,16 @@ def post(self):
             len(references))
+
+
+class FeedbackMessageClear(base.BaseHandler):
```
Usually we end all the handler names with "Handler". Maybe ClearFeedbackMessageHandler?

seanlip replied:
Actually, better: FeedbackThreadViewEventHandler

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,28 @@ def pop_feedback_message_references(user_id, references_length):
 
 
+def clear_feedback_message_references(user_id, exploration_id, thread_id):
+    """Removes feedback message references associated with exploration."""
```
exploration --> a feedback thread?

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,28 @@ def pop_feedback_message_references(user_id, references_length):
+
+    updated_references = []
+    for reference in model.feedback_message_references:
+        if (exploration_id not in reference.values() or
```
This use of reference.values() seems a bit weird. Could you not compare it more directly?

prasanna08 replied:
So you mean like this
```
reference['exploration_id'] != exploration_id or reference['thread_id'] != thread_id
```
Sure

seanlip replied:
Yep -- it's always better to be more explicit, generally speaking.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,28 @@ def pop_feedback_message_references(user_id, references_length):
+            updated_references.append(reference)
+
+    if not updated_references:
+        # Allow task to process even if model is deleted. Email will not be
```
The call to model.delete() doesn't actually do anything with the task, so maybe a better docstring would be:

Note that any tasks remaining in the email queue will still be processed, but if the model for the given user does not exist, no email will be sent.

seanlip replied:
Also, can you add an additional note here regarding the scenario you mentioned in the main thread?

# Note that, since the task in the queue is not deleted, the following scenario may occur: ...

seanlip commented on core/templates/dev/head/exploration_editor/feedback_tab/ThreadDataService.js:
```diff
@@ -27,6 +27,7 @@ oppia.factory('threadDataService', [
   var _SUGGESTION_ACTION_HANDLER_URL = '/suggestionactionhandler/' +
     _expId + '/';
   var _THREAD_HANDLER_PREFIX = '/threadhandler/' + _expId + '/';
+  var _FEEDBACK_MESSAGE_CLEAR = '/task/email/feedbackmessageclear';
```
I think that the "clearing" is an implementation detail. From the frontend's point of view, it wants to record that the user has seen this feedback message thread. So a better name might be: 

_FEEDBACK_THREAD_VIEW_EVENT_URL

(note that the 'URL' part is important, since it specifies what the constant actually is) and a corresponding url may be

/feedbackhandler/thread_view_event

similar to the ExplorationStartEventHandler etc. urls.

seanlip commented on feconf.py:
```diff
@@ -426,6 +426,8 @@ def get_empty_ratings():
 FEEDBACK_THREADLIST_URL_PREFIX = '/threadlisthandler'
 FEEDBACK_MESSAGE_EMAIL_HANDLER_URL = (
     '%s/feedbackemailhandler' % EMAILS_TASK_PREFIX)
+FEEDBACK_MESSAGE_EMAIL_CLEAR_URL = (
```
FEEDBACK_THREAD_VIEW_EVENT_URL?

seanlip replied:
(Also I don't think this has anything to do with the task itself, so no need for EMAILS_TASK_PREFIX)

seanlip commented:
Hi @prasanna08, this is looking good! And it's great that you identified the scenario above -- I incline to agree with you that it's fine to leave things as they are for now, since this case won't happen very often, but it might lead to some weird behaviour where an email is sent 1 minute (say) after a user gives feedback. I think we should leave a note in the codebase about this; I'll make a comment inline about where to put it.

Otherwise, a few comments, but they should be minor. Thanks!

prasanna08 commented:
Hi @seanlip, this commit addresses all your comments.
Thanks

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -357,6 +358,41 @@ def pop_feedback_message_references(user_id, references_length):
+        # Note that, since the task in the queue is not deleted, the following
+        # scenario may occur: If creator attends to arrived feedback bedore
+        # email is sent then model will be deleted but task will still execute
+        # after its countdown. Arrival new feedback arrives before task is
```
"Arrival new feedback arrives" --> "Arrival of new feedback"

seanlip commented on core/templates/dev/head/exploration_editor/feedback_tab/ThreadDataService.js:
```diff
@@ -69,6 +70,11 @@ oppia.factory('threadDataService', [
+      exploration_id: _expId,
+      thread_id: threadId
+    };
+    $http.post(_FEEDBACK_THREAD_VIEW_EVENT_URL, payload);
```
Actually, I just realized that this is misplaced (I think). Can it go into the controller function that is triggered when the user actually hits the button to see the new message? I'm concerned about this happening "behind the scenes" and marking stuff as read when it hasn't been.

seanlip replied:
Also, you can bring the payload inline to this call:

    $http.post(_FEEDBACK_THREAD_VIEW_EVENT_URL, {
      exploration_id: _expId,
      thread_id: threadId
    });

seanlip replied:
Or, actually -- if this is where we group all the HTTP calls, then a better design approach is to create a new function in the public API for this service that is something like markThreadAsSeen(..., ...). That function then makes the $http POST to the backend. The controller calls this function explicitly when the user clicks the button to open a thread.

prasanna08 replied:
Ok, I just checked. Actually function which called when thread is opened is `setActiveThread() in FeedbackTab.js`. This function makes call to `fetchMessages() function in ThreadDataServices.js` which then calls `_fetchMessages()` and currently I have put post request in implementation of this `_fetchMessages`. So where to actually put it, according to you ?
In `setActiveThread()` or in `fetchMessages()` ?

seanlip replied:
setActiveThread() should call threadDataServices.markThreadAsSeen(threadId).

seanlip commented:
LGTM, feel free to merge once Travis checks pass. Thanks!
